clients/drcachesim/tests/raw2trace_unit_tests.cpp

```diff
@@ -1974,6 +1974,312 @@ test_xfer_absolute(void *drcontext)
+        instr_t *nop3 = XINST_CREATE_nop(drcontext);
+        instr_t *move =
+            XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+        instr_t *store = XINST_CREATE_store(drcontext, OPND_CREATE_MEMPTR(REG2, 0),
+                                            opnd_create_reg(REG1));
+        instr_t *jcc =
+            XINST_CREATE_jump_cond(drcontext, DR_PRED_EQ, opnd_create_instr(store));
+        instrlist_append(ilist, nop1);
+        instrlist_append(ilist, jcc);
+        instrlist_append(ilist, nop2);
+        instrlist_append(ilist, nop3);
+        instrlist_append(ilist, store);
+        instrlist_append(ilist, move);
+        size_t offs_nop1 = 0;
+        size_t offs_jcc = offs_nop1 + instr_length(drcontext, nop1);
+        size_t offs_nop2 = offs_jcc + instr_length(drcontext, jcc);
+        size_t offs_nop3 = offs_nop2 + instr_length(drcontext, nop2);
+        size_t offs_store = offs_nop3 + instr_length(drcontext, nop3);
+        size_t offs_mov = offs_store + instr_length(drcontext, store);
+
+        std::vector<offline_entry_t> raw;
+        raw.push_back(make_header());
+        raw.push_back(make_tid());
+        raw.push_back(make_pid());
+        raw.push_back(make_line_size());
+        raw.push_back(make_marker(TRACE_MARKER_TYPE_RSEQ_ENTRY, offs_mov));
+        raw.push_back(make_block(offs_jcc, 1));
+        // The end of our rseq sequence, ending in a committing store.
+        raw.push_back(make_block(offs_store, 1));
+        raw.push_back(make_memref(42));
+        raw.push_back(make_marker(TRACE_MARKER_TYPE_RSEQ_ABORT, offs_mov));
+        raw.push_back(make_marker(TRACE_MARKER_TYPE_KERNEL_EVENT, offs_mov));
+        raw.push_back(make_block(offs_mov, 1));
+        raw.push_back(make_exit());
+
+        std::vector<trace_entry_t> entries;
+        if (!run_raw2trace(drcontext, raw, ilist, entries))
+            return false;
+        int idx = 0;
+        res = res && check_entry(entries, idx, TRACE_TYPE_HEADER, -1) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_VERSION) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_FILETYPE) &&
+            check_entry(entries, idx, TRACE_TYPE_THREAD, -1) &&
+            check_entry(entries, idx, TRACE_TYPE_PID, -1) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER,
+                        TRACE_MARKER_TYPE_CACHE_LINE_SIZE) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER,
+                        TRACE_MARKER_TYPE_CHUNK_INSTR_COUNT) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_RSEQ_ENTRY) &&
+            check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+#ifdef X86_32
+            // An extra encoding entry is needed.
+            check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+#endif
+            check_entry(entries, idx, TRACE_TYPE_INSTR_TAKEN_JUMP, -1, offs_jcc) &&
```

Comment:
Why is this branch marked taken? Is it because the next instr is a store which is the target of the jcc? But since the rseq end was rolled back, can we even tell whether that branch was really taken?

Reply:
Yes because the the branch targeted the store and we always roll back just the store.  Yes we can tell; you can tell in the final trace even without the new _TAKEN_ type but it takes some inference comparing to the rseq bounds and knowing how we roll back.

Reply:
This case, hit by a real user, is one of the driving forces behind adding the TAKEN labels, because for this rseq case you can't just look at the next record and figure out whether taken from just that.